src/main.rs

```diff
@@ -470,26 +476,34 @@ fn start_vmm(cmd_arguments: ArgMatches, api_socket_path: &str) -> Result<(), Err
         .map_err(Error::VmmThread)
 }
 
-fn main() {
-    // Ensure all created files (.e.g sockets) are only accessible by this user
-    let _ = unsafe { libc::umask(0o077) };
-
+fn default_api_server_path() -> Option<String> {
     let pid = unsafe { libc::getpid() };
     let uid = unsafe { libc::getuid() };
 
-    let mut api_server_path = format! {"/run/user/{}/cloud-hypervisor.{}", uid, pid};
-    if uid == 0 {
+    let api_server_directory = if uid == 0 {
         // If we're running as root, we try to get the real user ID if we've been sudo'ed
         // or else create our socket directly under /run.
         let key = "SUDO_UID";
         match env::var(key) {
-            Ok(sudo_uid) => {
-                api_server_path = format! {"/run/user/{}/cloud-hypervisor.{}", sudo_uid, pid}
-            }
-            Err(_) => api_server_path = format! {"/run/cloud-hypervisor.{}", pid},
+            Ok(sudo_uid) => format!("/run/user/{}", sudo_uid),
+            Err(_) => "/run".to_owned(),
         }
+    } else {
+        format!("/run/user/{}", uid)
+    };
+
+    if std::path::Path::new(&api_server_directory).exists() {
+        Some(format!("{}/cloud-hypervisor.{}", api_server_directory, pid))
```

Comment:
If someone provides their own path, shouldn't we fail if the path does not exist?  Otherwise we are silently failing correct and they won't know why?

Reply:
Well this code was only for the default case but i've removed it now. The updated PR neatly handles the user giving an invalid path by hoisting the creation of the socket out from the thread and propagating its error.